### PR TITLE
internal/charmstore: copy mgo sessions

### PR DESCRIPTION
**Pull request by rogpeppe**
We introduce a new Pool type which cannot be used for
charmstore operations, forcing clients to call Store to
operate on it.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/896/
Test PASSed.

File: internal/charmstore/hash.go [hunk 1]
```diff
@@ -35,7 +35,9 @@ func (s *Store) UpdateEntitySHA256(id *router.ResolvedURL) (string, error) {
 	// Update the entry asynchronously because it doesn't matter if it succeeds
 	// or fails, or if several instances of the charm store do it concurrently,
 	// and it doesn't need to be on the critical path for API endpoints.
-	go UpdateEntitySHA256(s, id, sum256)
+	s.Go(func(s *Store) {
```
Comment (mhilton) [inline internal/charmstore/hash.go:38]: nice

File: internal/charmstore/stats.go [hunk 2]
```diff
@@ -131,13 +141,13 @@ func (s *Store) statsTokenId(token string) (id int, found bool) {
...
 	if found {
-		s.cacheStatsTokenId(token, id)
+		s.cacheTokenId(token, id)
 	}
 	return
 }
 
-// statsIdToken returns the token for id from the cache, if found.
-func (s *Store) statsIdToken(id int) (token string, found bool) {
+// idToken returns the token for id from the cache, if found.
```
Comment (mhilton) [inline internal/charmstore/stats.go:149]: thanks for the drive by renaming

File: internal/charmstore/stats.go [hunk 3]
```diff
@@ -161,10 +171,21 @@ func timeToStamp(t time.Time) int32 {
...
 func (s *Store) IncCounterAsync(key []string) {
-	go func() {
+	s.Go(func(s *Store) {
 		if err := s.IncCounter(key); err != nil {
 			logger.Errorf("cannot increase stats counter for key %v: %v", key, err)
 		}
+	})
+}
+
+// Go runs the given function in a new goroutine,
```
Comment (mhilton) [inline internal/charmstore/stats.go]: should this be in stats.go? wouldn't store.go make more sense?

**Comment by mhilton** [comment]
:+1: this is great, thanks

**Comment by urosj** [comment]
:+1: With a next PR that deals with shorter timeout on /debug/status and /debug/info

File: internal/charmstore/stats.go (hunk 3 above)
Comment (rogpeppe) [inline internal/charmstore/stats.go]: good call. done.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/897/
Test PASSed.

File: internal/charmstore/stats.go [hunk 4]
```diff
@@ -161,10 +171,21 @@ func timeToStamp(t time.Time) int32 {
...
+// passing it a copy of s, which will be closed
+// after the function returns.
```
Comment (jrwren) [inline internal/charmstore/stats.go]: I was going to ask "why the s.Go(...) calls?"  Now I see the benefit. Really nice.

**Pull request merged**